knowrob_srdl/src/urdf_parser.cpp

```diff
@@ -0,0 +1,852 @@
+                PL_A3 = "mesh";
+                return true;
+            }
+            default:
+                return false;
+        }
+    } catch (const std::runtime_error& e) {
+        ROS_ERROR("%s", e.what());
+        return false;
+    }
+}
+
+PREDICATE(link_collision_name, 3) {
+    try {
+        std::string link_name((char*) PL_A1);
+        long index = (long) PL_A2;
+        urdf::LinkConstSharedPtr link = get_link(link_name);
+        if (!link_has_collision_with_index(link, index) ||
+                (link->collision_array[index]->name.compare("") == 0))
+            return false;
+        PL_A3 = link->collision_array[index]->name.c_str();
+        return true;
+    } catch (const std::runtime_error& e) {
+        ROS_ERROR("%s", e.what());
+        return false;
+    }
+}
+
+PREDICATE(link_collision_origin, 3) {
+    try {
+        std::string link_name((char*) PL_A1);
+        long index = (long) PL_A2;
+        urdf::LinkConstSharedPtr link = get_link(link_name);
+        if (!link_has_collision_with_index(link, index))
+            return false;
+        PL_A3 = to_prolog_pose(link->collision_array[index]->origin);
+        return true;
+    } catch (const std::runtime_error& e) {
+        ROS_ERROR("%s", e.what());
+        return false;
+    }
+}
+
+PREDICATE(link_collision_geometry, 3) {
+    try {
+        std::string link_name((char*) PL_A1);
+        long index = (long) PL_A2;
+        urdf::LinkConstSharedPtr link = get_link(link_name);
+        if (!link_has_collision_with_index(link, index) ||
+                !link->collision_array[index]->geometry)
+            return false;
+        PL_A3 = to_prolog_geometry(link->collision_array[index]->geometry);
+        return true;
+    } catch (const std::runtime_error& e) {
+        ROS_ERROR("%s", e.what());
+        return false;
+    }
+}
+
+/**************************************/
+/******** JOINT PROPERTIES ************/
+/**************************************/
+
+urdf::JointConstSharedPtr get_joint(const std::string& joint_name) {
+    urdf::JointConstSharedPtr joint = get_robot_model()->getJoint(joint_name);
+    if (!joint)
+        throw std::runtime_error("No joint with name '" + joint_name + "' in parsed URDF.");
+    return joint;
+}
+
+bool joint_has_pos_limits(urdf::JointConstSharedPtr joint) {
+    return joint->type == urdf::Joint::PRISMATIC || joint->type == urdf::Joint::REVOLUTE;
+}
+
+bool joint_has_vel_limit(urdf::JointConstSharedPtr joint) {
+    return joint->type == urdf::Joint::REVOLUTE ||
+           joint->type == urdf::Joint::PRISMATIC ||
+           joint->type == urdf::Joint::CONTINUOUS;
+}
+
+bool joint_has_effort_limit(urdf::JointConstSharedPtr joint) {
+    // joints that have velocity limits should also have effort limits
+    return joint_has_vel_limit(joint);
+}
+
+
+PREDICATE(joint_names, 1) {
+    try {
+        PlTail names(PL_A1);
+        for (auto const& joint_entry: get_robot_model()->joints_)
```

Comment:
never seen this keyword... so the type is inferred by the compiler, or what does it mean?